mws/utils/params.py

```diff
@@ -0,0 +1,231 @@
+
+    cleaned_params = dict()
+    for key, val in params.items():
+        try:
+            cleaned_params[key] = clean_value(val)
+        except ValueError as exc:
+            raise MWSError(str(exc)) from exc
+    return cleaned_params
+
+
+def clean_value(val):
+    """Attempts to clean a value so that it can be sent in a request."""
+    if isinstance(val, (dict, list, set, tuple)):
+        raise ValueError("Cannot clean parameter value of type %s" % str(type(val)))
+
+    if isinstance(val, (datetime.datetime, datetime.date)):
+        return clean_date(val)
+    if isinstance(val, bool):
+        return clean_bool(val)
+
+    # For all else, assume a string, and clean that.
+    return clean_string(str(val))
+
+
+def clean_string(val):
+    """Passes a string value through `urllib.parse.quote` to clean it.
+
+    Safe characters permitted: -_.~
+    """
+    return quote(val, safe="-_.~")
+
+
+def clean_bool(val):
+    """Converts a boolean value to its JSON string equivalent."""
+    if val is not True and val is not False:
+        raise ValueError("Expected a boolean, got %s" % val)
+    return json.dumps(val)
+
+
+def clean_date(val):
+    """Converts a datetime.datetime or datetime.date to ISO 8601 string.
+    Further passes that string through `urllib.parse.quote`.
+    """
+    return clean_string(val.isoformat())
+
+
+def flat_param_dict(value, prefix=""):
+    """Returns a flattened params dictionary by collapsing nested dicts and
+    non-string iterables.
+
+    Any arbitrarily-nested dict or iterable will be expanded and flattened.
+
+    - Each key in a child dict will be concatenated to its parent key.
+    - Elements of a non-string iterable will be enumerated using a 1-based index,
+      with the index number concatenated to the parent key.
+    - In both cases, keys and sub-keys are joined by ``.``.
+
+    If ``prefix`` is set, all keys in the resulting output will begin with
+    ``prefix + '.'``.
+    """
+    prefix = "" if not prefix else str(prefix)
+    # Prefix is now either an empty string or a valid prefix string ending in '.'
+    # NOTE should ensure that a `None` value is changed to empty string, as well.
+
+    if isinstance(value, str) or not isinstance(value, (Mapping, Iterable)):
+        # Value is not one of the types we want to expand.
+        if prefix:
+            # Can return a single dict of the prefix and value as a base case
+            if prefix.endswith("."):
+                # Remove an extraneous '.' at the end of this base case.
+                prefix = prefix[:-1]
+            return {prefix: value}
+        raise ValueError(
+            (
+                "Non-dict, non-iterable value requires a prefix "
+                "(would return a mapping of `prefix: value`)"
+            )
+        )
+
+    # Past here, the value is something that must be expanded.
+    # We'll build that output with recursive calls to `flat_param_dict`.
+
+    if prefix and not prefix.endswith("."):
```

Comment:
this is repetition of adding `.` to the end, worth making it a helper function of its own?

Reply:
Resolved, moved this to new utility `dot_appended_param` that DRYs up the process.